### PR TITLE
Issue #12, Tests for relationship

### PR DESCRIPTION
**Pull request by andrii-z4i**

### CODE REVIEW

File: xmind/tests/test_relationship_element.py [hunk 1]
```diff
@@ -0,0 +1,320 @@
...
+        self._add_attribute.assert_called_with(_obj1, "id")
+
+        _obj2 = RelationshipElement(1)
+        self._wb_mixin_element_init.assert_called_with(_obj2, 1, None)
+        self.assertEqual(_obj2.TAG_NAME, "relationship")
+        self._add_attribute.assert_called_with(_obj2, "id")
+
+        _obj3 = RelationshipElement(1, 2)
+        self._wb_mixin_element_init.assert_called_with(_obj3, 1, 2)
+        self.assertEqual(_obj3.TAG_NAME, "relationship")
```
Comment (mykola-pysarev) [inline xmind/tests/test_relationship_element.py]: as discussed, should be value imported from const.py instead of hardcoded value

Comment (mykola-pysarev) [reply xmind/tests/test_relationship_element.py]: this is related both to the "relationship" and "id"

File: xmind/tests/test_relationship_element.py [hunk 2]
```diff
@@ -0,0 +1,320 @@
...
+
+        with patch.object(_relationship_element, 'getFirstChildNodeByTagName') as _gf:
+            _gf.return_value = 'Something'
+            self.assertEqual('Something', _relationship_element._get_title())
+
+        self._wb_mixin_element_init.assert_called_once_with(
+            _relationship_element, 12, None)
+        self._add_attribute.assert_called_once_with(
+            _relationship_element, "id")
+        _gf.assert_called_once_with("title")
```
Comment (mykola-pysarev) [inline xmind/tests/test_relationship_element.py]: same about const values

File: xmind/tests/test_relationship_element.py (hunk 1 above)
Comment (andrii-z4i) [reply xmind/tests/test_relationship_element.py]: addressed

File: xmind/tests/test_relationship_element.py (hunk 2 above)
Comment (andrii-z4i) [reply xmind/tests/test_relationship_element.py]: addressed

File: xmind/tests/test_relationship_element.py [hunk 3]
```diff
@@ -0,0 +1,324 @@
...
+            _relationship_element, 12, None)
+        self._add_attribute.assert_called_once_with(
+            _relationship_element, ATTR_ID)
+        _wb_mock.getElementById.assert_called_once_with('1')
+        _tag_name.assert_called_once()
+
+    def test_find_end_point_endpoint_tag_name_is_topic(self):
+        _wb_mock = MagicMock()
+        _end_point = MagicMock()
+        _tag_name = PropertyMock(return_value='topic')
```
Comment (mykola-pysarev) [inline xmind/tests/test_relationship_element.py]: seems like the 'topic is also from const.py

File: xmind/tests/test_relationship_element.py [hunk 4]
```diff
@@ -0,0 +1,324 @@
...
+
+        with patch.object(_relationship_element, 'getAttribute') as _get_attribute:
+            _get_attribute.return_value = 120
+            self.assertEqual(120, _relationship_element.getEnd1ID())
+
+        self._wb_mixin_element_init.assert_called_once_with(
+            _relationship_element, 12, None)
+        self._add_attribute.assert_called_once_with(
+            _relationship_element, ATTR_ID)
+        _get_attribute.assert_called_once_with("end1")
```
Comment (mykola-pysarev) [inline xmind/tests/test_relationship_element.py]: "end1"  => const.ATTR_END1

File: xmind/tests/test_relationship_element.py [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+    def test_setEnd1ID(self):
+        _relationship_element = RelationshipElement(12)
+        _set_attribute = patch.object(
+            _relationship_element, 'setAttribute').start()
+        _update_modified_time = patch.object(
+            _relationship_element, 'updateModifiedTime').start()
+
+        self.assertIsNone(_relationship_element.setEnd1ID(15))
+
+        _set_attribute.assert_called_once_with("end1", 15)
```
Comment (mykola-pysarev) [inline xmind/tests/test_relationship_element.py]: "end1"=> const.ATTR_END1

File: xmind/tests/test_relationship_element.py [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+
+        with patch.object(_relationship_element, 'getAttribute') as _get_attribute:
+            _get_attribute.return_value = 120
+            self.assertEqual(120, _relationship_element.getEnd2ID())
+
+        self._wb_mixin_element_init.assert_called_once_with(
+            _relationship_element, 12, None)
+        self._add_attribute.assert_called_once_with(
+            _relationship_element, ATTR_ID)
+        _get_attribute.assert_called_once_with("end2")
```
Comment (mykola-pysarev) [inline xmind/tests/test_relationship_element.py]: same thing

File: xmind/tests/test_relationship_element.py [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    def test_setEnd2ID(self):
+        _relationship_element = RelationshipElement(12)
+        _set_attribute = patch.object(
+            _relationship_element, 'setAttribute').start()
+        _update_modified_time = patch.object(
+            _relationship_element, 'updateModifiedTime').start()
+
+        self.assertIsNone(_relationship_element.setEnd2ID(17))
+
+        _set_attribute.assert_called_once_with("end2", 17)
```
Comment (mykola-pysarev) [inline xmind/tests/test_relationship_element.py]: same

File: xmind/tests/test_relationship_element.py [hunk 8]
```diff
@@ -0,0 +1,324 @@
...
+        _title_mock = Mock()
+        _title_mock.getTextContent.return_value = 'it works'
+        _title_element = patch('xmind.core.relationship.TitleElement').start()
+        _title_element.return_value = _title_mock
+        _get_owner_workbook = patch.object(
+            _relationship_element, 'getOwnerWorkbook').start()
+        _get_owner_workbook.return_value = 'somehow'
+
+        with patch.object(_relationship_element, '_get_title') as _get_title:
+            _get_title.return_value = 'None'
```
Comment (mykola-pysarev) [inline xmind/tests/test_relationship_element.py]: 'None' - as for me its a bad value, I have spent 5 min for invstigation before i understood that this is a string value - not None

File: xmind/tests/test_relationship_element.py [hunk 9]
```diff
@@ -0,0 +1,324 @@
...
+        _get_owner_workbook = patch.object(
+            _relationship_element, 'getOwnerWorkbook').start()
+        _get_owner_workbook.return_value = 'somehow'
+        _append_child = patch.object(
+            _relationship_element, 'appendChild').start()
+        _update_modified_time = patch.object(
+            _relationship_element, 'updateModifiedTime').start()
+
+        with patch.object(_relationship_element, '_get_title') as _get_title:
+            _get_title.return_value = 'None'
```
Comment (mykola-pysarev) [inline xmind/tests/test_relationship_element.py]: same thing about 'None'

**Review by andrii-z4i — COMMENTED** [review]
@mykola-pysarev , @vlad-shulika 

Attention!!!
-----

All review comments are addressed
Please revisit it.

**Review by mykola-pysarev — APPROVED** [review]

**Pull request merged**